Source/CombatExtended/Harmony/Harmony_AttackTargetFinder.cs

```diff
@@ -49,6 +99,68 @@ internal static void Prefix(IAttackTargetSearcher searcher)
                                                .Select(t => t.TryGetComp<CompProjectileInterceptor>())
                                                .ToList() ?? new List<CompProjectileInterceptor>();
             }
+
+            /// <summary>
+            /// Find the best target for a ranged attacker (which may require the attacker to change their shooting position).
+            /// </summary>
+            /// <remarks>
+            /// This is an optimized equivalent of the vanilla logic in <see cref="AttackTargetFinder.BestAttackTarget"/>.
+            /// </remarks>
+            private static IAttackTarget FindAttackTargetForRangedAttack(
+                TargetScanFlags targetScanFlags,
+                IAttackTargetSearcher searcher,
+                Predicate<IAttackTarget> attackTargetValidator,
+                float maxDist,
+                bool canBashDoors,
+                bool canBashFences
+            )
+            {
+                validTargets.Clear();
+
+                var searcherThing = searcher.Thing;
+                var potentialAttackTargets = searcherThing.Map.attackTargetsCache.GetPotentialTargetsFor(searcher);
```

Comment:
This can probably be further improved by using CE's thing tracker to only get pawns within range, rather than getting all hostile pawns and checking their distance one at a time.  

It's up to you if you want to do it as part of this PR, we can also do it after merging this PR.  I can open a PR against your branch if you want me to do it as part of this PR.